proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,90 @@
+# DXIL Scalarization
+
+* Proposal: [NNNN](NNNN-DXIL-Scalarization.md)
+* Author(s): [Farzon Lotfi](https://github.com/farzonl)
+* Sponsor: [Farzon Lotfi](https://github.com/farzonl)
+* Status: **Under Consideration**
+* Impacted Projects: Clang
+
+## Introduction
+As mentioned in [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#vectors) 
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+ we need to be able to scalarize datastructures, call instructions, and vector 
+ operations like mathamatical, logical and bitcasts. The goal of this proposal 
+ is to present a solution for LLVM's pass manager.
+
+## Motivation
+Without Scalarizing the data structures and call instructions we can't generate
+ valid DXIL.
+
+ ## Background
+ In DXC Scalarization and SROA are merged in unfortnate ways. Most of that code
+ can be found here: [`SROA_Parameter_HLSL` Module pass in `ScalarReplAggregatesHLSL.cpp`](https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp#L4263). 
+The specific instructions DXC operates on are `CallInst`, `LoadInst`, 
+`StoreInst`, `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, and 
+`MemIntrinsic`. Not covered in this proposal is the handling of `Allocas`.
+That will be addressed in a later proposal covering DXIL Legalization.
+
+## Proposal
+Our goal should be legal DIXL with behavioral similarity to DXC. To that end it
+ makes the most sense to use the Scalarizer pass.
+- [Scalarizer.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Scalarizer.cpp)
+
+The `scalarizer` pass with the `-scalarize-load-store` and 
+`-scalarize-variable-insert-extract` flags  will handle the `BitCastInst`, 
+`CallInst`, `LoadInst` and `StoreInst`, and `GetElementPtrInst` cases. The 
+`AddrSpaceCastInst` case isn't relevant for upstream LLVM.
+
+`MemIntrinsic` is the odd case out. We don't currently emit memcpy for
+the same cases we do in DXC.
+In many cases operations like memcpy and memset seem to happen in global scope.
+See Example:
+- https://hlsl.godbolt.org/z/8vx1Yee7b
```

Comment:
I'm not entirely sure what I'm meant to be looking at on this godbolt link. Are the clang ones likely to remain stable (eg, after you've done this work will the clang links no longer make the point they're trying to make now?)

Reply:
I can remove the godbolt link. Yes after we have done the work the godbolt like might no longer be valid as the clang portion is on trunk.

 The point I was trying to make is that we should wait to handle the `MemIntrinsic` cases until we have cbuffer support. If I were to figure it out today it would be a direct translation. But I imagine there might be some intermediary we want because the other buffer cases happen after DXIL lowering and we are doing scalarization before.